File: gof-design-patterns/cs/P00_01_Encapsulation/BankAccount.c
#include <stdio.h>
#include <stdlib.h>

// 不完全型でbalanceを隠蔽（カプセル化）
typedef struct BankAccount BankAccount;

// 関数ポインタ型の定義
typedef void (*DepositFunc)(BankAccount* self, double amount);
typedef void (*WithdrawFunc)(BankAccount* self, double amount);
typedef double (*GetBalanceFunc)(BankAccount* self);
typedef void (*DestroyFunc)(BankAccount* self);

// BankAccount構造体の定義
// 実際のデータメンバーは実装ファイル内で定義（カプセル化）
struct BankAccount {
    double balance;  // データを隠蔽（通常は実装ファイル内で定義）
    
    // 関数ポインタ（メソッドに相当）
    DepositFunc deposit;
    WithdrawFunc withdraw;
    GetBalanceFunc getBalance;
    DestroyFunc destroy;
};

// メソッドの実装
static void deposit_impl(BankAccount* self, double amount) {
    if (self != NULL && amount > 0) {
        self->balance += amount;
    }
}

static void withdraw_impl(BankAccount* self, double amount) {
    if (self != NULL && amount > 0 && amount <= self->balance) {
        self->balance -= amount;
    }
}

static double getBalance_impl(BankAccount* self) {
    if (self != NULL) {
        return self->balance;
    }
    return 0.0;
}

static void destroy_impl(BankAccount* self) {
    if (self != NULL) {
        free(self);
    }
}

// コンストラクタ（オブジェクト生成関数）
BankAccount* BankAccount_new(double initialBalance) {
    BankAccount* account = (BankAccount*)malloc(sizeof(BankAccount));
    if (account == NULL) {
        return NULL;
    }
    
    // データメンバーの初期化
    account->balance = initialBalance;
    
    // 関数ポインタの初期化（メソッドの割り当て）
    account->deposit = deposit_impl;
    account->withdraw = withdraw_impl;
    account->getBalance = getBalance_impl;
    account->destroy = destroy_impl;
    
    return account;
}

// main関数（テスト用）
int main(void) {
    // オブジェクトの作成
    BankAccount* account = BankAccount_new(1000);
    
    if (account == NULL) {
        fprintf(stderr, "メモリの割り当てに失敗しました\n");
        return 1;
    }
    
    // メソッドの呼び出し（関数ポインタ経由）
    account->deposit(account, 500);
    account->withdraw(account, 200);
    
    printf("残高: %.2f\n", account->getBalance(account));
    
    // メモリの解放
    account->destroy(account);
    
    return 0;
}

File: gof-design-patterns/pythons/P00_01_Encapsulation/BankAccount.py
class BankAccount:
    # その前に、そもそもPython の self って何？
    # java だと、self なんて渡さなくても、this で自分のメンバにアクセスできたけど
    # C言語での疑似実装に近い感じか？
    def __init__(self, initial_balance: float):
        """コンストラクタ：初期残高を設定"""
        # python では、宣言していい変数は、始めて利用した時に定義される。
        # python では、変数名の先頭に _ を付けると、private な変数として扱う。
        # __balance は private な変数として扱う
        # つまり、このインストラクタでの初期設定が、プライベート変数の宣言＋初期化となる
        self.__balance = initial_balance  # データをprivateで隠蔽（名前修飾を使用）
    
    def deposit(self, amount: float) -> None:
        """預金メソッド"""
        if amount > 0:
            self.__balance += amount
    
    def withdraw(self, amount: float) -> None:
        """引き出しメソッド"""
        if amount > 0 and amount <= self.__balance:
            self.__balance -= amount
    
    def get_balance(self) -> float:
        """残高を取得するメソッド"""
        return self.__balance


if __name__ == "__main__":
    account = BankAccount(1000)
    
    # 【通常の呼び出し】インスタンスから呼び出す（selfは自動的に渡される）
    account.deposit(500)
    account.withdraw(200)
    print(f"残高: {account.get_balance()}")
    
    # 【実はこれと同じ】クラスオブジェクト経由で呼び出す場合、明示的にselfを渡す必要がある
    # これにより、Pythonのメソッドが実際には関数であることが分かる
    print("\n=== メソッドバインディングの仕組みを確認 ===")
    print(f"account.deposit の型: {type(account.deposit)}")
    print(f"BankAccount.deposit の型: {type(BankAccount.deposit)}")
    
    # クラスオブジェクト経由で直接呼び出す場合（関数として扱う）
    BankAccount.deposit(account, 100)  # 明示的にself（accountインスタンス）を渡す
    print(f"クラスオブジェクト経由で呼び出し後の残高: {account.get_balance()}")
    
    # インスタンスから呼び出す場合（バウンドメソッドとして扱う）
    account.deposit(50)  # selfは自動的に渡される
    print(f"インスタンスから呼び出し後の残高: {account.get_balance()}")
    
    # 【参考】メソッドの実体を確認
    print("\n=== メソッドの実体 ===")
    print(f"account.deposit: {account.deposit}")
    print(f"BankAccount.deposit: {BankAccount.deposit}")
    print(f"account.deposit.__self__: {account.deposit.__self__}")  # バウンドされているインスタンス
    print(f"account.deposit.__func__: {account.deposit.__func__}")  # 元の関数
    
    #  Pythonの__は完全なプライベートではなく、名前修飾による「意図しないアクセスを避ける」仕組みです。
    # 外部から直接アクセスしてもエラーにはなりませんが、クラス内の属性とは別物として扱われます。
    # name mangling という仕組みらしい。
    print("\n=== 名前修飾（name mangling）の確認 ===")
    account.__balance = 100000  # これは別の属性として作成される
    print(f"残高（get_balance経由）: {account.get_balance()}")  # 元の__balanceは変更されていない
    print(f"残高（直接アクセス）: {account.__balance}")  # 新しく作成された属性
